### ISSUE TITLE
microdom getElementsByTagName

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10607]
|<img alt="int's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| int reported|
|-|-|
|Trac ID|trac#909|
|Type|defect|
|Created|2005-02-28 21:37:37Z|

Attachments:

* [microdom.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed6/ed665fd661589a7a1e89af479e198fcea0048c15/a5a776971740d494a6d8512d691e9f5cb0f82d2b.patch) (2333 bytes) - added by exarkun on 2008-06-06 18:43:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__909 909
type__defect defect
reporter__int int
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1109626657000000 1109626657000000
changetime__1212965439000000 1212965439000000
version__ 
owner__ 
cc__glyph cc__exarkun cc__spiv cc__int
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="int's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|int commented|
|-|-|

```
#!html
<pre>
.getElementsByTagName doesn't work as expected
for example:
=========================================
from twisted.web import microdom
data = '&lt;div>&lt;div>111&lt;/div>&lt;div>222&lt;/div>&lt;/div>'
doc_microdom = microdom.parseString(data)
node = doc_microdom.getElementsByTagName('div')[0]
print node.toxml()
node = node.getElementsByTagName('div')[0]
print node.toxml()
=========================================
Output:
=========================================
&lt;div>&lt;div>111&lt;/div>&lt;div>222&lt;/div>&lt;/div>
&lt;div>&lt;div>111&lt;/div>&lt;div>222&lt;/div>&lt;/div>
=========================================
But should be:
=========================================
&lt;div>&lt;div>111&lt;/div>&lt;div>222&lt;/div>&lt;/div>
&lt;div>111&lt;/div>
=========================================

twisted 1.3.0
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Should it?  I see no docstrings that specify either behaviour.  Why do you think
ignoring the root element of the current subtree is a better behaviour?

In fact, the test suite explicitly tests for the current behaviour:

    def testSearch(self):
        s = "&lt;foo>&lt;bar id='me' />&lt;baz>&lt;foo />&lt;/baz>&lt;/foo>"
        ...
        d = microdom.parseString(s)
        ...
        root = d.documentElement
        ...
        self.assertEquals(d.getElementsByTagName("foo"),
                          [root, root.lastChild().firstChild()])
        ...
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="int's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|int commented|
|-|-|

```
#!html
<pre>
http://www.w3.org/TR/2000/REC-DOM-Level-2-Core-20001113/core.html#ID-745549614

====
getElementsByTagName 
Returns a NodeList of all descendant Elements with a given tag name, in the
order in which they are encountered in a preorder traversal of this Element tree.

.........

descendant 
A descendant node of any node A is any node below A in a tree model of a
document, where "above" means "toward the root."
====
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Ok, fair enough.  Thanks for the link.

This is almost certainly going to cause backwards compat problems -- the test
suite relies on the broken behaviour, and so other code probably does too :(
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Microdom's stated goal is to be a stable DOM implementation, not to be a
standards-compliant one.  It was written at a time when the stability of the XML
parsers in the stdlib was a serious problem, breaking applications from one
version to the next.

If we have unit tests for bad behavior, my opinion is, let's keep the bad
behavior, and you can use a different DOM implementation if you want it to be
correct.

If someone else wants to write a patch that introduces a "compliant" flag to
parse or something, which defaults to False, I won't complain, but I would be +0
on that.  Unless there is some reason that DOM applications actually want to be
portable from/to microdom?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="int's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|int commented|
|-|-|

```
#!html
<pre>
I think such bad behavior should be changed for 2.0 release, because if the
standart exists, i am sure microdom must to be compliant.
It's to difficult to guess why your script doesn't work as expected
(as i expected after work with other standart-compliant implementations)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
It's not a regression from previous releases.  I see no particular urgency to
fix this for 2.0, especially if no patch is forthcoming. ;)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

How do we feel about this?

Do we still think the current behavior should be retained?

At the very least, the function should have unit tests and a docstring, even if the old behavior is going to be retained.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

been broken forever, might as well stay broken forever.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

Maybe the `StringType` and `UnicodeType` could be replaced by str and unicode. There are also a couple of has_key around that could be removed. But that's not really important, merge when you want.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [23913]) Apply microdom.patch to add documentation on the non-standard behavior of getElementsByTagName

Author: exarkun
Reviewer: therve
Fixes: [#10607](https://github.com/twisted/twisted/issues/10607)

Add a docstring to `getElementsByTagName` describing its behavior, including
the non-standard behavior that it can return the node it is passed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
